reference/src/functions/macros.md

```diff
@@ -0,0 +1,680 @@
+        i = i + 1;
+    };
+    result
+}
+```
+
+The `$` is there to indicate that these do not behave like their normal, non-macro counterparts. For
+type parameters, they can be instantiated with any type (even a reference type `&` or `&mut`), and
+they will satisfy any constraint. Similarly for parameters, they will not be evaluated eagerly, and
+instead the argument expression will be substituted at each usage.
+
+## Lambdas
+
+Lambdas a new type of expression that can be used only with `macro`s. These are used to pass code
+from the caller into the body of the `macro`. While the substition is done at compile time, they are
+used similarly to [anonymous functions](https://en.wikipedia.org/wiki/Anonymous_function),
+[lambdas](https://en.wikipedia.org/wiki/Lambda_calculus), or
+[closures](<https://en.wikipedia.org/wiki/Closure_(computer_programming)>) in other languages.
+
+As seen in the example above (`$f: |$T| -> $U`), lambdas types defined with the syntax
+
+```text
+|<type>,*| (-> <type>)?
+```
+
+A few examples
+
+```move
+|u64, u64| -> u128 // a lambda that takes two u64s and returns a u128
+|&mut vector<u8>| -> &mut u8 // a lambda that takes a &mut vector<u8> and returns a &mut u8
+```
+
+If the return type is not annotated, it is unit `()` by default.
+
+```move
+// the following are euiqvalent
+|&mut vector<u8>, u64|
+|&mut vector<u8>, u64| -> ()
+```
+
+Lambda expressions then are defined at the call site of the `macro` the syntax
+
+```text
+|(<identifier> (: <type>)?),*| <expression>
+|(<identifier> (: <type>)?),*| -> <type> { <expression> }
+```
+
+Note that if the return type is annotated, the body of the lambda must be enclosed in `{}`.
+
+Using the `map` macro defined above
+
+```move
+let v = vector[1, 2, 3];
+let doubled: vector<u64> = map!(v, |x| 2 * x);
+let bytes: vector<vector<u8>> = map!(v, |x| std::bcs::to_bytes(&x));
+```
+
+And with type annotations
+
+```move
+let doubled: vector<u64> = map!(v, |x: u64| 2 * x); // return type annotation optional
+let bytes: vector<vector<u8>> = map!(v, |x: u64| -> vector<u8> { std::bcs::to_bytes(&x) });
+```
+
+### Capturing
+
+Lambda expressions can also refer to variables in the scope where the lambda is defined. This is
+sometimes called "capturing".
+
+```move
+let res = foo();
+let incremented = map!(vector[1, 2, 3], |x| x + res);
+```
+
+Any variable can be captured, including mutable and immutable references.
+
+See the [Examples](#iterating-over-a-vector) section for more complicated usages.
+
+### Limitations
+
+Currently, lambdas can only be used directly in the call of a `macro` function. They cannot be bound
+to a variable.
+
+For example, the following is allowed
+
+```move
+let doubled: vector<u64> = map!(vector[1, 2, 3], |x| 2 * x);
+```
+
+But binding `|x| 2 * x` to a variable is not allowed
+
+```move
+let f = |x| 2 * x;
+//      ^^^^^^^^^ Error! Lambdas must be used directly in 'macro' calls
+let doubled: vector<u64> = map!(vector[1, 2, 3], f);
+```
```

Comment:
```suggestion
to a variable. For example, the following is code will produce an error:

```move
let f = |x| 2 * x;
//      ^^^^^^^^^ Error! Lambdas must be used directly in 'macro' calls
let doubled: vector<u64> = map!(vector[1, 2, 3], f);
```
```

Due to the examples above, I think this is sufficient.